### PR TITLE
[FLINK-20528][python] Add Python building blocks to make sure the basic functionality of Stream Group Table Aggregation could work

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will Add Python building blocks to make sure the basic functionality of Stream Group Table Aggregation could work*

## Verifying this change

This change added tests and can be verified as follows:

  - *IT case `test_flat_aggregate`*
  - *IT case `test_flat_aggregate_list_view`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/table/udf.py [hunk 1]
```diff
@@ -647,3 +692,72 @@ def udaf(f: Union[Callable, UserDefinedFunction, Type] = None,
     else:
         return _create_udaf(f, input_types, result_type, accumulator_type, func_type,
                             deterministic, name)
+
+
+def udtaf(f: Union[Callable, UserDefinedFunction, Type] = None,
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:697]:
```suggestion
def udtaf(f: Union[Callable, TableAggregateFunction, Type] = None,
```

File: flink-python/pyflink/table/udf.py [hunk 2]
```diff
@@ -647,3 +692,72 @@ def udaf(f: Union[Callable, UserDefinedFunction, Type] = None,
...
+                          same parameters. (default True)
+    :param name: the function name.
+    :param func_type: the type of the python function, available value: general
+                     (default: general)
+    :return: UserDefinedAggregateFunctionWrapper or function.
+
+    .. versionadded:: 1.13.0
+    """
+    if func_type != 'general':
+        raise ValueError("The func_type must be one of 'general', got %s."
```
Comment (dianfu) [inline flink-python/pyflink/table/udf.py:755]:
```suggestion
        raise ValueError("The func_type must be 'general', got %s."
```

Comment (dianfu) [reply flink-python/pyflink/table/udf.py:755]: So it still doesn't support pandas? Could you create a ticket?

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/PythonStreamGroupTableAggregateOperator.java [hunk 3]
```diff
@@ -39,30 +37,19 @@
...
 
-	private final PythonAggregateFunctionInfo aggregateFunction;
-
-	private final DataViewUtils.DataViewSpec[] dataViewSpecs;
-
 	public PythonStreamGroupTableAggregateOperator(
 		Configuration config,
 		RowType inputType,
 		RowType outputType,
-		PythonAggregateFunctionInfo aggregateFunction,
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/PythonStreamGroupTableAggregateOperator.java:50]: What's the purpose of this change?

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks a lot for the PR. Have left a few comments.

File: flink-python/pyflink/table/udf.py (hunk 2 above)
Comment (HuangXingBo) [reply flink-python/pyflink/table/udf.py:755]: Yes. Unbounded stream will not support Pandas TableAggregateFunction, but group window stream can support Pandas TableAggregateFunction.

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/PythonStreamGroupTableAggregateOperator.java (hunk 3 above)
Comment (HuangXingBo) [reply flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/PythonStreamGroupTableAggregateOperator.java:50]: When there are multiple flat_maps, an additional PythonAggregateFunctionInfo of count(*) will be added to handle the situation of retract.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the update. I have addressed the comments at the latest commit.

**Review by dianfu — APPROVED** [review]
